src/Nethermind/Nethermind.Evm/EvmObjectFormat.cs

```diff
@@ -0,0 +1,208 @@
+            }
+        }
+        private int? _codeEndOffset = default;
+        public int CodeEndOffset
+        {
+            get
+            {
+                if (_codeEndOffset is null)
+                    _codeEndOffset = CodeStartOffset + CodeSize;
+                return _codeEndOffset.Value;
+            }
+        }
+        #endregion
+    }
+
+
+    public class EvmObjectFormat
+    {
+        private readonly ILogger _logger;
+        private bool LoggingEnabled => _logger is not null;
+        public EvmObjectFormat(ILogger logger = null)
+            => _logger = logger;
+
+        // magic prefix : EofFormatByte is the first byte, EofFormatDiff is chosen to diff from previously rejected contract according to EIP3541
+        private const byte EofMagicLength = 2;
+        private const byte EofFormatByte = 0xEF;
+        private const byte EofFormatDiff = 0x00;
+        private byte[] EofMagic => new byte[] { EofFormatByte, EofFormatDiff };
+
+        public bool HasEOFFormat(Span<byte> code) => code.Length > EofMagicLength && code.StartsWith(EofMagic);
+        public bool ExtractHeader(Span<byte> code, out EofHeader header)
+        {
+            if (!HasEOFFormat(code))
+            {
+                if (LoggingEnabled && _logger.IsTrace)
+                {
+                    _logger.Trace($"EIP-3540 : Code doesn't start with Magic byte sequence expected {EofMagic.ToHexString(true)} ");
+                }
+                header = null; return false;
+            }
+
+            int codeLen = code.Length;
+
+            int i = EofMagicLength;
+            int EOFVersion = code[i++];
+            switch (EOFVersion)
+            {
+                case 1:
+                    return HandleEOF1(code, out header, codeLen, ref i);
+                default:
+                    if (LoggingEnabled && _logger.IsTrace)
+                    {
+                        _logger.Trace($"EIP-3540 : Code has wrong EOFn version expected {1} but found {code[i]}");
+                    }
+                    header = null; return false;
+            }
+        }
+
+        private bool HandleEOF1(Span<byte> code, out EofHeader header, int codeLen, ref int i)
+        {
+            bool continueParsing = true;
+            header = new EofHeader
+            {
+            };
+
+            while (i < codeLen && continueParsing)
+            {
+                var sectionKind = (SectionDividor)code[i];
+                i++;
+
+                switch (sectionKind)
+                {
+                    case SectionDividor.Terminator:
+                        {
+                            continueParsing = false;
+                            break;
+                        }
+                    case SectionDividor.CodeSection:
+                        {
+                            // code-section must come first and there can be only one data-section
+                            if (header.CodeSize > 0 || i + 2 > codeLen)
```

Comment:
2 -> magic number?

Reply:
Each SectionSIze is 2 bytes long, I will separate the checks, and include the suitable Log messages